contracts/AdExRecoveryDAO.sol

```diff
@@ -0,0 +1,157 @@
+// SPDX-License-Identifier: agpl-3.0
+
+pragma solidity ^0.6.12;
+pragma experimental ABIEncoderV2;
+
+import "./Identity.sol";
+
+library RecoveryRequestLibrary {
+    struct RecoveryRequest {
+        address identity;
+        address newWalletAddress;
+    }
+
+    function hash(RecoveryRequest memory request)
+        internal
+        view
+        returns (bytes32)
+    {
+        return keccak256(abi.encode(
+            address(this),
+            request.identity,
+            request.newWalletAddress
+        ));
+    }
+}
+
+contract AdExRecoveryDAO {
+    using SafeMath for uint256;
+    using RecoveryRequestLibrary for RecoveryRequestLibrary.RecoveryRequest;
+
+    address public adminAddr;
+    uint256 public immutable MINIMUM_RECOVERY_DELAY; // in seconds
+    uint256 public recoveryDelay; // in seconds
+
+    mapping(address => bool) public proposers;
+    // recovery request hash to finalize time
+    mapping(bytes32 => uint256) public recovery;
+
+    event LogAddProposer(address proposer, uint256 timestamp);
+    event LogRemoveProposer(address proposer, uint256 timestamp);
+    event LogProposeRecovery(bytes32 recoveryId, address proposer, address identity, uint256 timestamp);
+    event LogFinalizeRecovery(bytes32 recoveryId, address proposer, address identity, uint256 timestamp);
+    event LogCancelRecovery(bytes32 recoveryId, address identity, uint256 timestamp);
+    event LogChangeAdmin(address oldAmin, address newAdmin);
+    event LogChangeRecoveryDelay(uint256 previousDelay, uint256 newDelay);
+
+    constructor(address admin, uint256 minDelay, uint256 delay) public {
+        require(delay > 0, 'INVALID_DELAY');
+        require(minDelay > 0, 'INVALID_MIN_DELAY');
+        require(admin != address(0), 'INVALID_ADMIN');
+
+        adminAddr = admin; 
+        recoveryDelay = delay;
+        MINIMUM_RECOVERY_DELAY = minDelay;
+    }
+
+    /**
+     * @notice Only the admin is allowed to add a proposer
+     * @param proposer The address of the proposer
+     */
+    function addProposer(address proposer) external {
+        require(msg.sender == adminAddr, 'ONLY_ADMIN_CAN_ADD_PROPOSER');
+        proposers[proposer] = true;
+        emit LogAddProposer(proposer, now);
+    }
+
+    /**
+     * @notice Only the admin is allowed to remove a proposer
+     * @param proposer The address of the proposer
+     */
+    function removeProposer(address proposer) external {
+        require(msg.sender == adminAddr, 'ONLY_ADMIN_CAN_REMOVE_PROPOSER');
+        delete proposers[proposer];
+        emit LogRemoveProposer(proposer, now);
+    }
+
+    /**
+     * @notice Only proposers are allowed to a propose recovery request
+     * @param request The details of the recovery request
+     */
+    function proposeRecovery(RecoveryRequestLibrary.RecoveryRequest memory request) external {
+        require(proposers[msg.sender] == true, 'ONLY_WHITELISTED_PROPOSER');
```

Comment:
same for this require the one that is used below in `finalizeRecovery`. Btw this sends "ONLY_WHITELISTED_PROPOSER" the other one is "ONLY_WHITELISTED_PROPOSERS".